core/interop/src/loaders/embedded.rs

```diff
@@ -0,0 +1,163 @@
+//! Embedded module loader. Creates a `ModuleLoader` instance that contains
+//! files embedded in the binary at build time.
+
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::path::Path;
+
+use boa_engine::module::{ModuleLoader, Referrer};
+use boa_engine::{Context, JsError, JsNativeError, JsResult, JsString, Module, Source};
+
+/// Normalize a specifier to remove `.` and `..` components.
+fn normalize_specifier(specifier: &JsString) -> JsResult<JsString> {
+    let specifier = specifier.to_std_string_escaped();
+    let components = specifier.split('/').collect::<Vec<_>>();
+    let specifier = components
+        .into_iter()
+        .try_fold(String::new(), |mut acc, component| {
+            if component == "." {
+                return Ok(acc);
+            }
+
+            if component == ".." {
+                if acc.is_empty() {
+                    return Err(JsError::from_native(
+                        JsNativeError::typ().with_message("invalid specifier".to_string()),
+                    ));
+                }
+
+                acc.pop();
+                return Ok(acc);
+            }
+
+            if !acc.is_empty() {
+                acc.push('/');
+            }
+            acc.push_str(component);
+            Ok(acc)
+        })?;
+
+    Ok(specifier.into())
```

Comment:
I think you can parse directly to `Path` in this case, which should allow using something like cargo's `normalize_path` impl (https://github.com/rust-lang/cargo/blob/fede83ccf973457de319ba6fa0e36ead454d2e20/src/cargo/util/paths.rs#L61)

Reply:
We should actually have a NewType for `JsSpecifier` that matches the semantics of `Path`. But I'm too lazy. In this case I cannot use `path` as for Windows `/`s are not going to be understood. That's also why I didn't just transform into `Path` elsewhere when doing resolution.

Reply:
> We should actually have a NewType for `JsSpecifier` that matches the semantics of `Path`

I don't think we can have that, because some embedders will want to use URLs as specifiers, so that's probably discarded.

> In this case I cannot use path as for Windows /s are not going to be understood. That's also why I didn't just transform into Path elsewhere when doing resolution.

Would this be solved if you used a strict `UnixPath` type, such as the one offered by [`typed-path`](https://crates.io/crates/typed-path)?

Reply:
> URLs as specifiers, so that's probably discarded.

A `JsSpecifier` should cover everything that the JavaScript specification says can go in an import specifier. Then allow to convert to other types.

Reply:
> A JsSpecifier should cover everything that the JavaScript specification says can go in an import specifier. Then allow to convert to other types.

When I say "cover the semantics of Path" what I mean is being able to iterate over components and such. Specifiers in JavaScript are, AFAIK, a subset of UTF-16 (can you put emojis or RTL/LTR controls in there?) and definitely a subset of OsStr.

Ah, I misunderstood. Yeah, that would be nice, maybe have some methods on there to convert from `JsSpecifier` to `JsPath`, `JsUrl`, etc.

> Maybe? I was trying to keep the number of dependencies low. If you're okay with it I can move to typed-path, it seems maintained.

Yep, I think it should be fine in this case because the crate is nicely maintained and it could possibly be added to the std library in the future.

Reply:
Adding a test, I found a bug, so I'm fixing that too. Was using the old `SimpleModuleLoader` that didn't take the referrer into account.